src/cmap/background.rs

```diff
@@ -1,71 +1,78 @@
 use std::{
-    sync::{atomic::Ordering, Weak},
     time::Duration,
+    sync::Weak,
 };
 
+use futures_timer::Delay;
+
 use super::{ConnectionPool, ConnectionPoolInner};
-use crate::event::cmap::ConnectionClosedReason;
+use crate::{RUNTIME, event::cmap::ConnectionClosedReason};
 
 /// Initializes the background thread for a connection pool. A weak reference is used to ensure that
 /// the connection pool is not kept alive by the background thread; the background thread will
 /// terminate if the weak reference cannot be converted to a strong reference.
-pub(crate) fn start_background_thread(pool: Weak<ConnectionPoolInner>) {
-    std::thread::spawn(move || loop {
-        match pool.upgrade() {
-            Some(pool) => perform_checks(pool.into()),
-            None => return,
-        };
+pub(crate) fn start_background_task(pool: Weak<ConnectionPoolInner>) {
+    RUNTIME.execute(async move {
+        loop {
+            match pool.upgrade() {
+                Some(pool) => perform_checks(pool.into()).await,
+                None => return,
+            };
 
-        std::thread::sleep(Duration::from_millis(10));
+            Delay::new(Duration::from_millis(10)).await;
+        }
     });
 }
 
 /// Cleans up any stale or idle connections and adds new connections if the total number is below
 /// the min pool size.
-fn perform_checks(pool: ConnectionPool) {
+async fn perform_checks(pool: ConnectionPool) {
     // We remove the perished connections first to ensure that the number of connections does not
     // dip under the min pool size due to the removals.
-    remove_perished_connections_from_pool(&pool);
-    ensure_min_connections_in_pool(&pool);
+    pool.remove_perished_connections().await;
+    pool.ensure_min_connections().await;
 }
 
-/// Iterate over the connections and remove any that are stale or idle.
-fn remove_perished_connections_from_pool(pool: &ConnectionPool) {
-    let mut connections = pool.inner.connections.write().unwrap();
-    let mut i = 0;
+impl ConnectionPool {
```

Comment:
these are all pretty much the same thing, but I moved them to be methods on `ConnectionPool` itself.